usb: Add USB OTG HS support for STM32H7S3 (NUCLEO-H7S3L8)

jw0785 opened the pull request:
Fixed:
- [x] match methods to the latest commit in upstream
- [x] add node zephyr_udc0: &usbotg_hs in board dts

Reference: 
- https://github.com/zephyrproject-rtos/zephyr/compare/main...FRASTM:zephyr:stm32h7rs_otg_hs

Test:
- [x] test passed
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/cdc_acm/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/cdc_acm_bridge/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/hid-keyboard/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/hid-mouse/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/midi/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/testusb/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/webusb
- [x] test failed
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/dfu*/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/uac2_explicit_feedback/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/uac2_implicit_feedback/
  - [x] west build -p always -b nucleo_h7s3l8 samples/subsys/usb/mass/

TODO:
- [ ] integrate [mcuboot](https://github.com/zephyrproject-rtos/mcuboot/compare/main...FRASTM:mcuboot:xip_h7s3)

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1075,6 +1075,21 @@ static int priv_clock_enable(void)
+	 * VDDUSB independent USB supply (PWR clock is on)
+	 * with LL_PWR_EnableVDDUSB function (higher case)
+	 */
+	LL_PWR_EnableVDDUSB();
```
This is already implemented at line 1075, and as it is, it seems not consistent with what you expect for H7RX platforms.
I think you should move your added sequence under an `#elif` directive  inside the previous `#if`/`#elif` series:
```patch
	HAL_SYSCFG_SetOTGPHYReferenceClockSelection(SYSCFG_OTG_HS_PHY_CLK_SELECT_1);
	/* Configuring the SYSCFG registers OTG_HS PHY : OTG_HS PHY enable*/
	HAL_SYSCFG_EnableOTGPHY(SYSCFG_OTG_HS_PHY_ENABLE);
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_otghs)
 	/* Enable Vdd USB voltage monitoring */
 	LL_PWR_EnableVddUSBMonitoring();
 	while (__HAL_PWR_GET_FLAG(PWR_FLAG_USB33RDY)) {
 		/* Wait FOR VDD33USB ready */
 	}
 	/* Enable VDDUSB */
 	LL_PWR_EnableVddUSB();
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) && defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	LL_PWR_EnableUSBReg();
+	LL_PWR_EnableUSBHSPHYReg();
+	LL_PWR_EnableUSBVoltageDetector();
+	__HAL_RCC_USB_OTG_HS_CLK_ENABLE();
+	/* Configuring the SYSCFG registers OTG_HS PHY : OTG_HS PHY enable*/
+	__HAL_RCC_USBPHYC_CLK_ENABLE();
 #elif defined(PWR_USBSCR_USB33SV) || defined(PWR_SVMCR_USV)
 	/*
 	 * VDDUSB independent USB supply (PWR clock is on)
 	 * with LL_PWR_EnableVDDUSB function (higher case)
 	 */
 	LL_PWR_EnableVDDUSB();
 #endif
```

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1098,7 +1113,10 @@ static int priv_clock_enable(void)
+	
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX) && DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
+	/* USB OTG HS receives the 60MHz from RCC : no domain clock expected */
+#else
```
I think this should better be handled from **drivers/usb/udc/Kconfig.stm32**:
```patch
 config UDC_STM32_CLOCK_CHECK
 	bool "Runtime USB 48MHz clock check"
-	default y if !(SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32U5X)
+	default y if !(SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32U5X || SOC_SERIES_STM32H7X)
 	help
 	  Enable USB clock 48MHz configuration runtime check.
 	  In specific cases, this check might provide wrong verdict and should
 	  be disabled.
```

etienne-lms reviewed (commented):
Comment for commit "drivers: usb: stm32 usb device controller correct USB PHY".

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32h7rs_clock.h:
```diff
@@ -75,7 +75,9 @@
-#define OTGFS_SEL(val)		STM32_DT_CLOCK_SELECT((val), 3, 14, D1CCIPR_REG)
+#define USBREF_SEL(val)		STM32_DT_CLOCK_SELECT(val, 15, 8, D1CCIPR_REG)
+#define USBPHYC_SEL(val)	STM32_DT_CLOCK_SELECT(val, 3, 12, D1CCIPR_REG)
+#define OTGFS_SEL(val)		STM32_DT_CLOCK_SELECT(val, 3, 14, D1CCIPR_REG)
```
Replace `val` with `(val)`

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -363,7 +371,9 @@ static int usb_dc_stm32_phy_specific_clock_enable(const struct device *const clk
-
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX) && DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
+	/* USB OTG HS receives the 60MHz from RCC : no domain clock expected */
+#else
```
Same comment as for **udc_stm32.c**: handle that from **drivers/usb/udc/Kconfig.stm32:**?

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -389,11 +389,9 @@ static int usb_dc_stm32_phy_specific_clock_enable(const struct device *const clk
 		}
 	}
-
+#endif
```
This fixup should be squash in commit "drivers: usb: stm32 device stack supporting the stm32h7rs" (if still applicable)

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -442,7 +440,7 @@ static int usb_dc_stm32_clock_enable(void)
 						LL_AHB2_GRP1_PERIPH_USBPHY);
 #elif !DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_otghs)
-	LL_AHB1_GRP1_DisableClockLowPower(LL_AHB1_GRP1_PERIPH_OTGHSULPI);
+	LL_AHB1_GRP1_DisableClockSleep(LL_AHB1_GRP1_PERIPH_USBOTGHS);
```
This N6 fixup deserves a specific commit.

jw0785 commented:
Hi, I will finish what i left in a couple of weeks, get preoccupied lately.

erwango commented on dts/arm/st/h7rs/stm32h7rs.dtsi:
```diff
@@ -329,6 +329,10 @@
 				clocks = <&rcc STM32_CLOCK(AHB4, 15U)>;
 			};
+
+			pinctrl_dummy: dummy_pins {
+				pinmux = <0>;
+			};
```
Why adding these dummy nodes ?
Also, if modifying the original commit, please update the commit message and add you sign-off. Or better, don't touch the commit and add a new commit

erwango replied:
> the reason being when trying west build, i find it requires dma-names and a non-empty pinctrl in sail_a and sail_b to build now.

If this is the case, then we should fix sai and make a dedicated PR for that.

erwango replied:
@mariopaja Do you see any reason that would be required ?
@jw0785 Can you provide more details as you seem to be the only one impacted ?

erwango commented on drivers/usb/device/Kconfig:
```diff
@@ -63,7 +63,7 @@ config USB_DC_STM32
 	bool "Runtime USB 48MHz clock check"
 	depends on USB_DC_STM32
-	default y if !(SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32U5X)
+	default y if !(SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32U5X || SOC_SERIES_STM32H7X || SOC_SERIES_STM32H7RSX)
```
See previous comment about modifying original commits.
Also, doing changes on SOC_SERIES_STM32H7X should be done using a dedicated commit and proper justification. In this case a dedicated PR would even be required

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -59,28 +60,28 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
+	 DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
 
-#define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
-			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+#define USB_OTG_HS_ULPI_PHY                                                                        \
+	(DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
```
Don't mix style change or file clean up with functional changes.
Make a dedicated commit with specific message

erwango commented on drivers/usb/udc/Kconfig.stm32:
```diff
@@ -37,7 +37,7 @@ config UDC_STM32_MAX_QMESSAGES
 config UDC_STM32_CLOCK_CHECK
 	bool "Runtime USB 48MHz clock check"
-	default y if !(SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32U5X)
+	default y if !(SOC_SERIES_STM32F1X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32U5X || SOC_SERIES_STM32H7X || SOC_SERIES_STM32H7RSX)
```
Ditto

erwango commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -27,58 +27,58 @@
 #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
-#define DT_DRV_COMPAT st_stm32_otghs
-#define UDC_STM32_IRQ_NAME     otghs
+#define DT_DRV_COMPAT      st_stm32_otghs
+#define UDC_STM32_IRQ_NAME otghs
```
Ditto

erwango commented on include/zephyr/dt-bindings/clock/stm32h7rs_clock.h:
```diff
@@ -17,116 +17,118 @@
 
 /** Fixed clocks  */
 /* Low speed clocks defined in stm32_common_clocks.h */
-#define STM32_SRC_HSE		(STM32_SRC_LSI + 1)
```
I don't see the changes described in the commit message, only clean up.

erwango reviewed (changes requested):
Please review all commits and align content with description.
Don't do modifications on other's commits without notice, or better, at all.
Do dedicated commits for clean up (if really clean up is required)

Commit:
style: add dummy nodes in stm32h7rs.dtsi

Add dummy nodes in stm32h7rs.dtsi to make sai nodes ci compilant

Signed-off-by: Wei Wang <wwng.work@gmail.com>

Commit:
dts: arm: stm32h7rs add the usb OTG HS nodes

Add the usb OTG HS nodes to the stm32h7rs series

Signed-off-by: Wei Wang <wwng.work@gmail.com>

Commit:
boards: st: nucleo_h7s318: enable usb OTG HS

Enable the usb OTG HS node on the nucleo_h7s3l8 board

Signed-off-by: Wei Wang <wwng.work@gmail.com>

Commit:
Kconfig: add kconfig entries for STM32H7RSX series

disable UDC_STM32_CLOCK_CHECK for STM32H7RSX

Signed-off-by: Wei Wang <wwng.work@gmail.com>

Commit:
drivers: usb: stm32 usb device controller correct USB PHY

Set the correct the phy_itface depending on the setting
Save few lines after resetting the priv->pcd structure
with the memset
based on original work by Francois Ramu

Signed-off-by: Wei Wang <wwng.work@gmail.com>

jfischer-no commented on drivers/usb/device/Kconfig:
We should not add new features to deprecated drivers or subsystems. These changes must be dropped.

jfischer-no commented on drivers/usb/device/usb_dc_stm32.c:
We should not add new features to deprecated drivers or subsystems. These changes must be dropped.

jw0785 commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -418,6 +426,8 @@ static int usb_dc_stm32_clock_enable(void)
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
 	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
+#elif defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	LL_AHB1_GRP1_DisableClockSleep(LL_AHB1_GRP1_PERIPH_USBOTGHS || LL_AHB1_GRP1_PERIPH_USBPHYC);
```
will
```c
LL_AHB1_GRP1_DisableClockSleep(LL_AHB1_GRP1_PERIPH_USBOTGHS);
```
without checking phy do?

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1088,6 +1088,13 @@ static int priv_clock_enable(void)
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) && defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	LL_PWR_EnableUSBReg();
+	LL_PWR_EnableUSBHSPHYReg();
+	LL_PWR_EnableUSBVoltageDetector();
```
This is required for operation in `otgfs` mode too. (arguably outside PR scope)

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1088,6 +1088,13 @@ static int priv_clock_enable(void)
+	LL_PWR_EnableUSBReg();
+	LL_PWR_EnableUSBHSPHYReg();
+	LL_PWR_EnableUSBVoltageDetector();
+	__HAL_RCC_USB_OTG_HS_CLK_ENABLE();
```
Already done by `clock_control_on()`
```suggestion
```

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1088,6 +1088,13 @@ static int priv_clock_enable(void)
+	LL_PWR_EnableUSBHSPHYReg();
+	LL_PWR_EnableUSBVoltageDetector();
+	__HAL_RCC_USB_OTG_HS_CLK_ENABLE();
+	/* Configuring the SYSCFG registers OTG_HS PHY : OTG_HS PHY enable*/
+	__HAL_RCC_USBPHYC_CLK_ENABLE();
```
```suggestion
	/* Enable OTG_HS PHY clock */
	__HAL_RCC_USBPHYC_CLK_ENABLE();
```

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1158,6 +1165,8 @@ static int priv_clock_enable(void)
 	 */
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
 	LL_AHB1_GRP1_DisableClockSleep(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
+#elif defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	LL_AHB1_GRP1_DisableClockSleep(LL_AHB1_GRP1_PERIPH_USBOTGHS);
```
I don't think this is correct.
